subsys/bluetooth/controller/ll_sw/ull.c

```diff
@@ -406,21 +406,34 @@ void ll_reset(void)
 
 	/* Reset LLL via mayfly */
 	{
-		u32_t retval;
-		struct k_sem sem;
 		static memq_link_t link;
 		static struct mayfly mfy = {0, 0, &link, NULL,
 					    perform_lll_reset};
+		u32_t retval;
+
+		/* NOTE: If Zero Latency Interrupt is used then LLL context
+		 *       will be higher priority hence completing inline
+		 *       in this function. Else use semaphore to wait for
+		 *       perform_lll_reset to complete.
+		 */
+
+#if !defined(CONFIG_BT_CTLR_ZLI)
+		struct k_sem sem;
 
 		k_sem_init(&sem, 0, 1);
 		mfy.param = &sem;
+#endif /* !CONFIG_BT_CTLR_ZLI */
+
 		retval = mayfly_enqueue(TICKER_USER_ID_THREAD,
 					TICKER_USER_ID_LLL, 0, &mfy);
 		LL_ASSERT(!retval);
+
+#if !defined(CONFIG_BT_CTLR_ZLI)
 		/* LLL reset must complete before returning - wait for
```

Comment:
How do you wait without the semaphore then?

Reply:
Good question. Is it done directly via mayfly_enqueue(), in a nested-fashion?

Reply:
The mayfly enqueue of callee, same or higher priority, by a caller translates to execution of callee by direct function call or pre-emption to callee ISR respectively. i.e. mayfly returns with the callee execution complete. There is no explicit wait required.

This is when mayfly backend/hal are using ISRs, which is what ZLI support applies to.

If ZLI is not used, then it can be threads/meta IRQs being used for mayfly backend/hal, in this case semaphores are used for wait.